### PR TITLE
jobscript: add lib/python to PYTHONPATH

### PR DESCRIPTION
**Pull request by oliver-sanders**
#### PR Template

This is a small change with no associated Issue.

Requirements check-list:
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] My commits have logically grouped changes (use interactive rebase
      to tidy your branch history if necessary).
- [x] I have not included off-topic changes (use other PRs for these).
- [x] I have included tests with adequate coverage.
- [x] I have added an entry in `CHANGES.md` for next release.

#### The Change

Small change to add `<suitedir>/lib/python` into PYTHONPATH in the task execution environment.

The motive here is to make it easier for users to link the task scripts in the suite bin directory to suite-specific Python libraries. Currently the user is responsible for doing the PYTHONPATH melding.

#### Potential Caveats

1. Users could do something stupid like create the file `<suitedir>/lib/python/os.py`
2. We are now mixing lots of different libraries into the same `lib/python` dir.
3. The `CYLC_SUITE_RUN_DIR` / `CYLC_SUITE_DEF_PATH` arrangement is horrible, this will get resolved with this with the migration of `rose suite-run` functionality?

### CODE REVIEW

**Comment by kinow** [comment]
I thought I had seen some code while reviewing the cylc dir change that did exactly this.

**Comment by kinow** [comment]
Travis failed with

```
./tests/authentication/05-full-control.t      (Wstat: 256 Tests: 1 Failed: 0)
```

So just kicked it again to get a green build here :+1:

**Comment by kinow** [comment]
@oliver-sanders not sure if relevant or helpful, but here's the code I mentioned in another comment above:

https://github.com/cylc/cylc-flow/blob/b91594b10a7315191d25cd62d9fe29caa744217e/cylc/flow/scheduler.py#L143-L150

**Comment by oliver-sanders** [comment]
@kinow This is where the different ways Cylc spawns processes starts to mess with your head.

**tldr;** That code puts `lib/python` in the `PYTHONPATH` for the workflow process but that environment is not inherited by jobs.

Here are the different types of process and what *i think* happens (corrections welcome):

* **Subprocesses** spawned by the workflow process should share the same environment as the workflow process. (e.g. cylc event handlers).
* **Forked processes** inherit the same environment as the parent (e.g. the workflow process when daemonised).
* **Batch Jobs** are shell processes, they run in a login shell and may submit / run locally or remotely.
  * The `slurm` batch submission system makes a copy of the environment at submission time which gets really strange because jobs are submitted locally rather than on a login node which gets really really strange if a user submits a job interactively because `slurm` copies the complete environment including `DISPLAY` into a non-interactive environment.
  * Other batch submission systems can also do this "remote submission" though I don't know what consequences there are to this.
* **Background Jobs** are executed in a shell in a subshell executed by the cylc workflow process (does that mess with your head enough?), the job will show up in `ps` as `nohup bash -c 'exec ...'`. This should make background jobs behave the same as jobs submitted by batch submission (i.e. no shared environment with workflow process).

> here's the code I mentioned in another comment above

That code modifies the `PYTHONPATH` of the Cylc workflow process. The purpose of this is so we can use `lib/python` modules in cylc event handlers and xtriggers.

The job execution environment isn't a subprocess of the Cylc workflow process so doesn't inherit this `PYTHONPATH` meaning that we have to set it manually. Note also that the job could be running remotely so the `PYTHONPATH` needs to point at the suite directory (in a relative sense). This is the same for `PATH` which is why the `bin/` directory is handled in the same way.

*I've probably made some mistake(s) in here somewhere...*

**Comment by kinow** [comment]
Ooohh. That makes sense. But will re-read it all with calm tomorrow morning while reading tge code too. Thanks for the detailed explanation!

**Comment by hjoliver** [comment]
@oliver-sanders 

> ...This should make background jobs behave the same as jobs submitted by batch submission (i.e. no shared environment with workflow process).

This is not actually true any more, although it should be, and I believe it used to be.  I have recently verified by experiment that local background jobs inherit the workflow server environment.  (On master).

**Comment by kinow** [comment]
I think I will need more coffee ☕️ to understand how it all works. But I will try running some experiments too and see if I understand and what seems to work in my environment.

**Comment by matthewrmshin** [comment]
Can easily prefix the command with `env -i` if we really need to run background jobs in a pure environment. But is there a reason why we need to do that?

**Comment by hjoliver** [comment]
> 
> 
> Can easily prefix the command with `env -i` if we really need to run background jobs in a pure environment. But is there a reason why we need to do that?

The reason I think that should be done, is to make local background jobs behave identically to batch jobs and remote background jobs (with respect to job environment).  This consistency would be useful - no (environment-related) surprises on moving from testing a suite with local background jobs, to using a proper batch system (and/or a remote job host.

**Comment by matthewrmshin** [comment]
Not sure that is true. Because local batch jobs can also inherit, depending on how you set up the batch system and the job. Remote ones can't, for obvious reasons.

**Comment by hjoliver** [comment]
@kinow my TLDR (variation on @oliver-sanders ): this PR about making user's Python modules (in [suite-config-dir]/lib/python) automatically available to task jobs, which (except currently for local background jobs as just discussed) do not inherit the workflow server program environment.  The workflow server program also needs the same, in order to import user's local xtrigger modules for instance (but that doesn't help task jobs, because that environment is not inherited by jobs).

File: tests/rnd/02-lib-python-in-job/suite.rc [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+[runtime]
+    [[foo]]
+        script = """
+            # check PYTHONPATH is set correctly
+            grep -q "${CYLC_SUITE_RUN_DIR}/lib/python" <<< "${PYTHONPATH}"
+
+            # run a toy example
+            python -c '
+            from pub import beer
```
Comment (hjoliver) [inline tests/rnd/02-lib-python-in-job/suite.rc]: :beers:

Comment (oliver-sanders) [reply tests/rnd/02-lib-python-in-job/suite.rc]: :beer:
